### ISSUE TITLE
组合训练噪声问题

### ISSUE THREAD

**Issue by 12313415354** [FALLANGELZOU/END2#3]
你好，我在测试的时候，代码照着原来的测试随机噪声，发现提取率只有total_acc92约，单独测一个噪声发现提取率只有
total_acc: 0.7263052126011217，看到END2-main\Tools\steganography\Noiser 这个文件下的，class Noiser(nn.Module):这个类中向前传播方法，发现，这里训练的也是随机噪声。layer = random.choice(self.models)，  是不是这里要改成组合呢。
    def forward(self, img, cover_img):
        layer = random.choice(self.models)
        self.latest_layer = layer
        img, cover_img = self._normalize(img, cover_img)
        self._check(img, cover_img)
        img, cover_img = layer.noise(img, cover_img)
        img, cover_img = self._denormalize(img, cover_img)
        return img, cover_img
        pass

**Comment by FALLANGELZOU** [comment]
提取率问题可能需要检查训练曲线，尽管我们在文中提出两种缓解过拟合现象的方法，但不幸的是，模型在后期训练中仍然会出现Encoder逐渐过拟合的现象（在我们的测试中，这个现象在之前的几篇工作中都出现了，暂时还没有一个比较好的解决方案）， 通过早停策略来选择最优模型。此外，提取检测率和视觉质量强相关，最好统一在一个条件下测试。

我们给出的噪声层的默认参数比实际测试的参数强度要强不少，请问您测试的单独噪声具体参数是什么样的。

我没太懂您“其中("Identity", None),是不是要删除呢” 是什么意思，测量随机噪声的时候，我们是加入了("Identity", None)的，并且在训练中也不建议去除，因为这会导致模型对数据的原始分布产生误判。

**Comment by 12313415354** [comment]
> 提取率问题可能需要检查训练曲线，尽管我们在文中提出两种缓解过拟合现象的方法，但不幸的是，模型在后期训练中仍然会出现Encoder逐渐过拟合的现象（在我们的测试中，这个现象在之前的几篇工作中都出现了，暂时还没有一个比较好的解决方案）， 通过早停策略来选择最优模型。此外，提取检测率和视觉质量强相关，最好统一在一个条件下测试。
> 
> 我们给出的噪声层的默认参数比实际测试的参数强度要强不少，请问您测试的单独噪声具体参数是什么样的。
> 
> 我没太懂您“其中（”Identity“， None），是不是要删除呢” 是什么意思，测量随机噪声的时候，我们是加入了（“Identity”， None）的，并且在训练中也不建议去除，因为这会导致模型对数据的原始分布产生误判。

你好，非常感谢你回答我问题。我详细与您说下我刚才遇到的问题。 我按照readme的要求训练了模型也是3090训练的，在测试不同类型的失真的时候，也就是您论文里面的表 2：针对不同失真的鲁棒性的基准比较。我尝试在修改self.noiser = Noiser([
("Identity", None),("Rotate", None),("Crop", None),("Translate", None),("Scale", None),("Shear", None),("Dropout", None),("Cropout", None),("Color", None),("KorniaJpeg", None),("GaussianFilter", None),("GaussianNoise", None)])  注释其他噪声选择单一变量，即分开测，让随机噪声成为特定的一种噪声，发现acc没达到表二的要求。我查看您的论文的时候说的训练的时候，是组合噪声吧，测试是随机噪声，我想着是不是可能我使用的随机噪声参与的训练呢？  于是我找到一下代码中 layer = random.choice(self.models)，认为训练的可能是随机噪声训练的。   
class Noiser(nn.Module):
    def __init__(self, params=[("Identity", None),], normalize=False):
        """噪声组合层，图像需要在[0,1]之间
        Args:
            params (list, optional): 噪声层参数, 格式[(名称，{参数})]. Defaults to [("Identity", None),].
            normalize (bool, optional): 是否自动归一化到[0, 1]. Defaults to False.
        """
        super().__init__()
        self.models = []
        self.normalize = normalize
        self.minn = 0
        self.maxx = 1
        for name, kwargs in params:
            if kwargs is None:
                self.models.append(NoiseLayerManager.get(name)())
            else:
                self.models.append(NoiseLayerManager.get(name)(**kwargs))
            pass
        
        self.models = nn.ModuleList(self.models)
        self.latest_layer = None
        pass

    def forward(self, img, cover_img):
        layer = random.choice(self.models)
        self.latest_layer = layer
        img, cover_img = self._normalize(img, cover_img)
        self._check(img, cover_img)
        img, cover_img = layer.noise(img, cover_img)
        img, cover_img = self._denormalize(img, cover_img)
        return img, cover_img
        pass

**Comment by FALLANGELZOU** [comment]
可能是我们文章的表述有些问题，我们训练的时候就是使用随机噪声而非组合噪声。

此外，能给出具体的测试代码吗

**Comment by 12313415354** [comment]
> 可能是我们文章的表述有些问题，我们训练的时候就是使用随机噪声而非组合噪声。
> 
> 此外，能给出具体的测试代码吗
测试代码的话，我用的您论文中的valid如下，已经将 wm_img = clip_psnr(wm_img, img, 35, over_clip=True)设置了35.
with torch.no_grad():
    for batch in tqdm(dataloader):
        img = batch['img'].to(device)
        msg = batch['msg'].to(device)
        wm_img = model.encoder(img, msg)
        wm_img = clip_psnr(wm_img, img, 35, over_clip=True)

        noised_wm_img, _ = model.normal_noiser(wm_img, img)
        t_predict_msg, _ = model.decoder_t(noised_wm_img)
        s_predict_msg, _ = model.decoder_s(noised_wm_img)
        #predict_msg = t_predict_msg
        #predict_msg = (t_predict_msg + s_predict_msg) / 2 # 这里三种情况差不多
        predict_msg = s_predict_msg
        psnr = PSNR(wm_img, img)
        ssim = SSIM(wm_img, img)
        acc = msg_acc(predict_msg, msg)
        total_psnr += psnr.item()
        total_acc += acc.item()
        total_ssim += ssim.item()
        n += 1
        pass

并且我在Noiser.py这个文件中，控制单一变量。	
self.noiser = Noiser([
			# ("Identity", None),
			#("Rotate", None),
			# ("Crop", None),
			#("Translate", None),
			# ("Scale", None),
			# ("Shear", None),
			#("Dropout", None),
			#("Cropout", None),
			#
			
			# ("Color", None),
			#("KorniaJpeg", {"min_q": 50, "max_q": 50})
			("GaussianFilter", None),
			# ("GaussianNoise", None)
		])
最后在相应的噪声文件中，例如Filter.py文件中，例如：sigma = 2.0强行设置（下面），设置相应的噪声）
def noise(self, img, cover_img, repeat=False):
		if repeat is False:
			self.now_kernel = np.random.randint(self.min_kernel, self.max_kernel)//2*2+1
			# self.now_sigma = np.random.uniform(self.min_sigma, self.max_sigma)
			self.now_sigma = 2.0
			print(self.now_kernel, self.now_sigma)
		img = gaussian_blur2d(img, (self.now_kernel, self.now_kernel), (self.now_sigma, self.now_sigma))
		return img, cover_img
		pass
最后运行valid.py文件结果是
total_psnr: 35.0
total_ssim: 0.9198577769341008
total_acc: 0.8346774097411863
我不知道我这样操作是否正确，请您指正。

**Comment by FALLANGELZOU** [comment]
看着没问题，感觉可能是训练出问题了，我们最近有点忙，这个代码提交的比较匆忙，后续我们check一下并提供对应的预训练权重

**Comment by 12313415354** [comment]
> 看着没问题，感觉可能是训练出问题了，我们最近有点忙，这个代码提交的比较匆忙，后续我们check一下并提供对应的预训练权重

非常感谢您的回复，期待您后面提供预训练权重。我会继续支持你们的工作的，感谢
